lib/Conversion/TorchToLinalg/Pooling.cpp

```diff
@@ -228,6 +228,149 @@ class ConvertAtenAdaptiveAvgPool2dOp
+      return rewriter.notifyMatchFailure(
+          op, "count_include_pad must be a constant");
+    } else if (!countIncludePad) {
+      return rewriter.notifyMatchFailure(
+          op, "unimplemented: count_include_pad is expected to be true");
+    }
+
+    // The divisor_override should be either None or a constant int.
+    int64_t divisorOverride;
+    if (!op.divisor_override().getType().isa<Torch::NoneType>() &&
+        (!matchPattern(op.divisor_override(),
+                       m_TorchConstantInt(&divisorOverride)))) {
+      return rewriter.notifyMatchFailure(
+          op, "divisor_override must be either None or a constant int");
+    }
+
+    SmallVector<int64_t, 4> paddingIncludingNC = {0, 0};
+    paddingIncludingNC.insert(paddingIncludingNC.end(), paddingInts.begin(),
+                              paddingInts.end());
+    Value paddedInput = torch_to_linalg::getPaddedTensor(op, rewriter, self,
+                                                         paddingIncludingNC);
+
+    Value N = getDimOp(rewriter, loc, self, 0);
+    Value C = getDimOp(rewriter, loc, self, 1);
+    Value H = getDimOp(rewriter, loc, self, 2);
+    Value W = getDimOp(rewriter, loc, self, 3);
+
+    SmallVector<Value> kernelSizeIntValues =
+        getAsConstantIntValues(rewriter, loc, kernelSizeInts);
+    SmallVector<Value> strideIntValues =
+        getAsConstantIntValues(rewriter, loc, strideInts);
+    SmallVector<Value> paddingIntValues =
+        getAsConstantIntValues(rewriter, loc, paddingInts);
+
+    Value cstOne =
+        rewriter.create<arith::ConstantOp>(loc, rewriter.getI64IntegerAttr(1));
+    Value Hout = torch_to_linalg::getOutputDimForConvOps(
+        rewriter, loc, H, paddingIntValues[0], /*dilation=*/cstOne,
+        kernelSizeIntValues[0], strideIntValues[0]);
+    Value Wout = torch_to_linalg::getOutputDimForConvOps(
+        rewriter, loc, W, paddingIntValues[1], /*dilation=*/cstOne,
+        kernelSizeIntValues[1], strideIntValues[1]);
+
+    // Initialize output tensor with smallest floating point value
+    Value outTensor = rewriter.create<linalg::InitTensorOp>(
+        loc, ValueRange{N, C, Hout, Wout}, elementType);
+    auto initialAttr = rewriter.getFloatAttr(
+        elementType,
+        APFloat::getSmallest(
+            elementType.cast<mlir::FloatType>().getFloatSemantics(),
+            /*Negative*/ true));
+    Value initValue = rewriter.create<arith::ConstantOp>(loc, initialAttr);
+    Value outTensorInitialized =
+        rewriter.create<linalg::FillOp>(loc, initValue, outTensor).getResult(0);
+
+    auto stridesAttr = rewriter.getI64VectorAttr(strideInts);
+    auto dilationAttr = rewriter.getI64VectorAttr({1, 1});
+    Value windowTensor = rewriter.create<linalg::InitTensorOp>(
+        loc, getAsConstantIndexValues(rewriter, loc, kernelSizeInts),
+        elementType);
+
+    Value sumPool2d = rewriter
+                          .create<linalg::PoolingNhwcSumOp>(
```

Comment:
The dimension layout required by the linalgOp doesn't match the torch op. We will need to add a new one upstream  https://github.com/llvm/llvm-project/blob/c9c2363048b00749561070008d3b696438ee1024/mlir/python/mlir/dialects/linalg/opdsl/ops/core_named_ops.py#L517

Reply:
Do you mean to say add a `linalg::PoolingNchwSumOp`? If yes, could you please tell me how to do that or give some references?

Reply:
@cathyzhyi Gentle reminder. I am waiting for your reply on this to go ahead.

Reply:
Sry, I was off yesterday. 
yea, we need to add a linalg.pooling_nchw_sum using Linalg opDSL. You could refer to this [doc page](https://mlir.llvm.org/docs/Dialects/Linalg/OpDSL/) and this PR as a reference https://reviews.llvm.org/D106658.

Reply:
I have added the linalg.pooling_nchw_sum op. https://reviews.llvm.org/D123013

Reply:
looks nice. let's wait for folks that are more familiar with this to review. Does the e2e test pass now with this new op?

Reply:
I haven't tried. I am waiting for it to get merged.

Reply:
> Does the e2e test pass now with this new op?

Yes.